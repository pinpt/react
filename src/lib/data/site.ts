import type { Site } from '../types/site';
import { executeAPI } from '../fetch';

export const fetchSite = async (siteId: string): Promise<Site> => {
	const { site } = await executeAPI(`/site-api/v1/site/${siteId}`);
	return site;
};

type AnalyticsResult = Record<string, { claps: number; pageviews: number }>;

export const fetchAnalytics = async (siteId: string): Promise<AnalyticsResult> => {
	const { data } = await executeAPI(`/site-api/v1/site/${siteId}/analytics`);
	return data;
};
