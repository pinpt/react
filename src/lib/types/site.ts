export interface ISite {
	id: string;
	logoUrl?: string;
	mailingAddress?: string;
	url: string; // the canonical url for the site
	slug: string;
	name: string;
	private: boolean; // true if private and should not be indexable by search engines
	theme?: {
		title?: string;
		social?: Record<string, string>;
		description?: string;
		logoLink?: string;
		copyright?: string;
		eulaLink?: string;
		privacyLink?: string;
		tosLink?: string;
		homepage?: {
			tags?: string[];
		};
	};
	scriptPlugins?: Record<string, any>;
	basePath: string;
}
