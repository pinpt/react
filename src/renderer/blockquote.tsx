import { registerNode, recurseIntoChildren, NodeProps } from './register';

const BlockQuote = ({ node }: NodeProps) => {
	return <blockquote>{recurseIntoChildren(node)}</blockquote>;
};

registerNode('blockquote', (node) => <BlockQuote node={node} />);
