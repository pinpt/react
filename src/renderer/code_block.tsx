import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { twilight } from 'react-syntax-highlighter/dist/cjs/styles/prism';
import { NodeProps, registerNode } from './register';

const CodeBlock = ({ node }: NodeProps) => {
	return (
		<SyntaxHighlighter
			language={node.attrs.language}
			style={twilight}
			customStyle={{
				background: 'black',
				fontFamily: 'ui-monospace',
				borderWidth: '0px',
				borderRadius: '.25em',
			}}
		>
			{node.content?.[0].text ?? ''}
		</SyntaxHighlighter>
	);
};

registerNode('code_block', (node) => <CodeBlock node={node} />);
