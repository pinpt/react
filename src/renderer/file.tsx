import {
	faDownload, faFile, faFileAlt, faFileArchive, faFileAudio, faFileCode, faFileExcel, faFilePdf,
	faFilePowerpoint, faFileVideo, faFileWord
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { NodeProps, registerNode } from './register';

export const iconForType = (type = '') => {
	if (/pdf/.test(type)) {
		return faFilePdf;
	}
	if (/zip|tar/.test(type)) {
		return faFileArchive;
	}
	if (/video/.test(type)) {
		return faFileVideo;
	}
	if (/audio/.test(type)) {
		return faFileAudio;
	}
	if (/excel|spreadsheet/.test(type)) {
		return faFileExcel;
	}
	if (/powerpoint/.test(type)) {
		return faFilePowerpoint;
	}
	if (/word/.test(type)) {
		return faFileWord;
	}
	if (/css|csv|graphql|html|java|json|php|python|typescript|xml/.test(type)) {
		return faFileCode;
	}
	if (/text/.test(type)) {
		return faFileAlt;
	}
	return faFile;
};

export const formattedSize = (size = 0) => {
	const kB = size / 1000;
	if (kB < 10) {
		return `${Math.round(kB * 10) / 10} KB`;
	}
	if (kB < 1000) {
		return `${Math.round(kB)} KB`;
	}
	const MB = kB / 1000;
	return `${Math.round(MB * 10) / 10} MB`;
};

const File = ({ node }: NodeProps) => {
	const { description, filename, publicUrl, size, type } = node.attrs;

	const onDownload = () => {
		if (publicUrl) {
			const u = new URL(publicUrl);
			u.searchParams.set('filename', filename);
			u.searchParams.set('disposition', 'attachment');
			const a = document.createElement('a');
			a.style.display = 'none';
			a.href = u.toString();
			// the filename you want
			a.download = filename;
			document.body.appendChild(a);
			a.click();
			a.remove();
		}
	};

	return (
		<div className="file">
			<div className="container">
				<FontAwesomeIcon icon={iconForType(type)} />
				<div className="details">
					{publicUrl ? (
						<a href={publicUrl} target="_blank" rel="noopener noreferrer" className="filename">
							{filename}
						</a>
					) : (
						<div className="filename">{filename}</div>
					)}
					{description && <div className="description">{description}</div>}
					<div className="size">{formattedSize(size)}</div>
				</div>
				<div className="options">
					{publicUrl && (
						<div className="option group" onClick={onDownload}>
							<FontAwesomeIcon icon={faDownload} />
							<span>Download</span>
						</div>
					)}
				</div>
			</div>
		</div>
	);
};

registerNode('file', (node) => <File node={node} />);
