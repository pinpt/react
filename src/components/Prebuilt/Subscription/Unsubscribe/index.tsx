import React, { ReactElement } from 'react';
import { ISite } from '../../../../lib/types/site';
import { ICopyrightProps } from '../../../Copyright';
import { IFooterProps } from '../../../Footer';
import { IHeaderProps } from '../../../Header';
import withWrapper from '../../../Internal/withWrapper';
import { ILogoProps } from '../../../Logo';
import { ISocialMediaBarProps } from '../../../SocialMedia/SocialMediaBar';
import UnsubscribeComponent from '../../../Subscription/Unsubscribe';
import { IThemeToggleProps } from '../../../ThemeToggle';
import Footer from '../../Footer';
import Header from '../../Header';

export interface IPrebuiltUnsubscribeProps {
	className?: string;
	site: ISite;
	renderHeader?: (site: ISite) => ReactElement<IHeaderProps>;
	handleSelectHome?: () => void;
	renderLogo?: (site: ISite) => ReactElement<ILogoProps>;
	renderThemeToggle?: (site: ISite) => ReactElement<IThemeToggleProps>;
	renderFooter?: (site: ISite) => ReactElement<IFooterProps>;
	renderSocial?: (site: ISite) => ReactElement<ISocialMediaBarProps>;
	renderCopyright?: (site: ISite) => ReactElement<ICopyrightProps>;
}

const Unsubscribe = (props: IPrebuiltUnsubscribeProps) => {
	const {
		className = '',
		renderHeader,
		site,
		handleSelectHome,
		renderLogo,
		renderThemeToggle,
		renderFooter,
		renderCopyright,
		renderSocial,
	} = props;

	return (
		<div className={`Pinpoint Page SubscriptionUnsubscribe Prebuilt ${className}`}>
			{withWrapper(
				renderHeader?.(site) ?? (
					<Header
						site={site}
						handleSelectHome={handleSelectHome}
						renderLogo={renderLogo}
						renderSearch={() => <></>}
						renderSubscribe={() => <></>}
						renderThemeToggle={renderThemeToggle}
					/>
				),
				'header'
			)}
			{withWrapper(<UnsubscribeComponent />, 'unsubscribe')}
			{withWrapper(
				renderFooter?.(site) ?? (
					<Footer
						site={site}
						renderCopyright={renderCopyright}
						renderLogo={renderLogo}
						renderSocial={renderSocial}
						renderSubscribe={() => <></>}
					/>
				),
				'footer'
			)}
		</div>
	);
};

export default Unsubscribe;
