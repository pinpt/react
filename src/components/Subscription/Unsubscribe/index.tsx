import React, { useCallback, useMemo, useState } from 'react';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { getFileUrl } from '../../../lib/file_metadata';
import useEmailAction from '../../../lib/hooks/useEmailAction';
import { getRouterAbsolutePath } from '../../../lib/router';
import { ISite } from '../../../lib/types/site';
import EmailAction from '../../EmailAction';

export interface IUnSubscribeProps {
	className?: string;
	logo?: string;
	name: string;
	subscribed?: boolean;
	email: string;
	handleUnsubscribe?: () => Promise<void>;
	handleSubscribe?: () => Promise<void>;
	manageSubscriptions?: () => void;
	showLogo?: boolean;
	site: ISite;
	pending?: boolean;
}

const baseClass = 'Pinpoint SubscriptionUnsubscribe';

const Unsubscribe = (props: IUnSubscribeProps) => {
	const {
		className = '',
		logo,
		name,
		subscribed = false,
		email,
		handleSubscribe,
		handleUnsubscribe,
		manageSubscriptions,
		showLogo = false,
		site,
		pending = false,
	} = props;
	const emailActionState = useEmailAction();

	const fileApi = useMemo(() => {
		if (getRouterAbsolutePath(site, '').includes('edge')) {
			return 'https://file.edge.pinpoint.com';
		}
	}, [site]);

	const onSubscribe = useCallback(async () => {
		try {
			emailActionState.setters.setMessage('');
			if (handleSubscribe) {
				await handleSubscribe();
				emailActionState.setters.setMessage(
					'You were successfully re-subscribed. Please check your email to verify your subscription.'
				);
			}
		} catch (ex: any) {
			emailActionState.setters.setCritical(false);
			emailActionState.setters.setError(ex.message);
		}
	}, [emailActionState.setters, handleSubscribe]);

	const onUnSubscribe = useCallback(async () => {
		try {
			emailActionState.setters.setMessage('');
			if (handleUnsubscribe) {
				await handleUnsubscribe();
				emailActionState.setters.setMessage(
					'You were successfully unsubscribed. Unsubscribed by accident? Re-subscribe below.'
				);
			}
		} catch (ex: any) {
			emailActionState.setters.setCritical(false);
			emailActionState.setters.setError(ex.message);
		}
	}, [emailActionState.setters, handleUnsubscribe]);

	return (
		<div className={`${baseClass} Wrapper ${className}`}>
			<div className={`${baseClass} Container`}>
				<div className={`${baseClass} Content`}>
					<EmailAction {...emailActionState} />
					{showLogo && (
						<>
							{logo && <img src={getFileUrl(logo, fileApi)} alt="" className={`${baseClass} Logo`} />}
							<h3 className={`${baseClass} Name`}>{name}</h3>
						</>
					)}
					<p className={`${baseClass} Warning`}>
						You are about to {subscribed ? 'unsubscribe' : 'subscribe'} <strong>{email}</strong>{' '}
						{subscribed ? 'from' : 'to'} updates from {name}.
					</p>
					<div className="pt-6">
						{subscribed ? (
							<button className={`${baseClass} Button Unsubscribe`} onClick={onUnSubscribe} disabled={pending}>
								{pending ? <FontAwesomeIcon icon={faSpinner} pulse /> : 'Unsubscribe'}
							</button>
						) : (
							<button className={`${baseClass} Button Resubscribe`} onClick={onSubscribe} disabled={pending}>
								{pending ? <FontAwesomeIcon icon={faSpinner} pulse /> : 'Resubscribe'}
							</button>
						)}
					</div>
					<span className={`${baseClass} Manage`} onClick={manageSubscriptions}>
						manage all subscriptions
					</span>
				</div>
			</div>
		</div>
	);
};

export default Unsubscribe;
