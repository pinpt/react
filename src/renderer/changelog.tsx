import { ReactYouTubeLite as Youtube } from 'react-youtube-lite';
import { getFileUrl } from '../lib/file_metadata';
import { slugifyContent } from '../lib/string';
import { ChangelogCoverMedia, ChangelogMediaType } from '../lib/types';
import { Content as Doc } from './';

const ImageMedia = ({ src, title = '' }: { src: string; title?: string }) => {
	return (
		<div className="image">
			<img src={src} alt={title} />
		</div>
	);
};

const VideoMedia = ({ src }: { src: string }) => {
	return (
		<div className="video">
			<video controls src={src} />
		</div>
	);
};

const YoutubeMediaSSR = ({ id, poster = 'hqdefault' }: { id: string; poster?: string }) => {
	// the component isn't SSR friendly so this is basically used to sort of mimic it for generated code
	const posterUrl = `https://i.ytimg.com/vi/${id}/${poster}.jpg`;
	return (
		<>
			{poster && <link rel="preload" href={posterUrl} as="image" />}
			<div
				className="youtube ryt-lite embed-responsive aspect-ratio-16/9"
				style={{ backgroundImage: `url(${posterUrl})` }}
			>
				<div className="lty-playbtn" data-url={`https://www.youtube-nocookie.com/embed/${id}?autoplay=1`} />
			</div>
		</>
	);
};

const YoutubeMedia = ({
	id,
	metadata,
	staticMode,
}: {
	id: string;
	metadata?: Record<string, any>;
	staticMode?: true;
}) => {
	return (
		<div className="youtube">
			{!staticMode ? (
				<Youtube url={`https://www.youtube.com/watch?v=${id}`} poster={metadata?.poster} />
			) : (
				<YoutubeMediaSSR id={id} poster={metadata?.poster} />
			)}
		</div>
	);
};

const Media = ({ media, title, staticMode }: { media?: ChangelogCoverMedia; title?: string; staticMode?: true }) => {
	if (!media) {
		return <></>;
	}
	let content: React.ReactElement;
	switch (media.type) {
		case ChangelogMediaType.None: {
			return <></>;
		}
		case ChangelogMediaType.Image: {
			content = <ImageMedia src={getFileUrl(media.value)} title={title} />;
			break;
		}
		case ChangelogMediaType.Video: {
			content = <VideoMedia src={getFileUrl(media.value)} />;
			break;
		}
		case ChangelogMediaType.Youtube: {
			content = <YoutubeMedia id={media.value} metadata={media.metadata} staticMode={staticMode} />;
			break;
		}
	}
	return <div className="media-container">{content}</div>;
};

interface ChangelogProps {
	id: string;
	title?: string;
	content?: any;
	coverMedia?: ChangelogCoverMedia;
	limit?: number;
	divider?: boolean;
	staticMode?: true;
}

const ChangelogLog = (props: ChangelogProps) => {
	return (
		<article className="changelog notebook-editor read-only">
			{props.coverMedia && (
				<section className="covermedia">
					<Media media={props.coverMedia} title={props.title} staticMode={props.staticMode} />
				</section>
			)}
			<section className="ProseMirror">
				<Doc node={props.content} limit={props.limit} />
				{props.limit && props.content.content?.length > props.limit && (
					<div className="continue">
						<a href={slugifyContent(props.id, props.title)}>Continue Reading â†’</a>
					</div>
				)}
			</section>
			{props.divider && <hr className="divider" />}
		</article>
	);
};

export default ChangelogLog;
