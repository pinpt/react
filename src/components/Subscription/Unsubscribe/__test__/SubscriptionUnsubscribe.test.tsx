import renderer from 'react-test-renderer';
import Unsubscribe from '..';
import site from '../../../Prebuilt/__data__/testSite.json';

test('Test default state', () => {
	const component = renderer.create(
		<Unsubscribe
			name="Pinpoint"
			logo="fileid:1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
			subscribed
			email="keegan@pinpoint.com"
			site={site}
		/>
	);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test pending state', () => {
	const component = renderer.create(
		<Unsubscribe
			name="Pinpoint"
			logo="fileid:1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
			subscribed
			email="keegan@pinpoint.com"
			site={site}
			pending
		/>
	);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test not subscribed state', () => {
	const component = renderer.create(
		<Unsubscribe
			name="Pinpoint"
			logo="fileid:1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
			email="keegan@pinpoint.com"
			site={site}
		/>
	);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test not subscribed pending state', () => {
	const component = renderer.create(
		<Unsubscribe
			name="Pinpoint"
			logo="fileid:1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
			email="keegan@pinpoint.com"
			site={site}
			pending
		/>
	);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
