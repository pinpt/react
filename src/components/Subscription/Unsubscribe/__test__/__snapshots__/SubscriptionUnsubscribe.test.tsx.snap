// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default state 1`] = `
<div
  className="Pinpoint SubscriptionUnsubscribe Wrapper "
>
  <div
    className="Pinpoint SubscriptionUnsubscribe Container"
  >
    <div
      className="Pinpoint SubscriptionUnsubscribe Content"
    >
      
      
      
      <p
        className="Pinpoint SubscriptionUnsubscribe Warning"
      >
        You are about to 
        unsubscribe
         
        <strong>
          keegan@pinpoint.com
        </strong>
         
        from
         updates from 
        Pinpoint
        .
      </p>
      <div
        className="pt-6"
      >
        <button
          className="Pinpoint SubscriptionUnsubscribe Button Unsubscribe"
          disabled={false}
          onClick={[Function]}
        >
          Unsubscribe
        </button>
      </div>
      <span
        className="Pinpoint SubscriptionUnsubscribe Manage"
      >
        manage all subscriptions
      </span>
    </div>
  </div>
</div>
`;

exports[`Test not subscribed state 1`] = `
<div
  className="Pinpoint SubscriptionUnsubscribe Wrapper "
>
  <div
    className="Pinpoint SubscriptionUnsubscribe Container"
  >
    <div
      className="Pinpoint SubscriptionUnsubscribe Content"
    >
      
      
      
      <p
        className="Pinpoint SubscriptionUnsubscribe Warning"
      >
        You are about to 
        subscribe
         
        <strong>
          keegan@pinpoint.com
        </strong>
         
        to
         updates from 
        Pinpoint
        .
      </p>
      <div
        className="pt-6"
      >
        <button
          className="Pinpoint SubscriptionUnsubscribe Button Resubscribe"
          disabled={false}
          onClick={[Function]}
        >
          Resubscribe
        </button>
      </div>
      <span
        className="Pinpoint SubscriptionUnsubscribe Manage"
      >
        manage all subscriptions
      </span>
    </div>
  </div>
</div>
`;
